### ISSUE TITLE
Flutter Engine Lerping Math is Broken for Certain Values

### ISSUE THREAD

**Issue by royzwerneman** [flutter/flutter#67423]
## Steps to Reproduce

1. Run [this gist](https://gist.github.com/royzwerneman/a21701355f15a9f2cddff0aa39816415) that uses problematic flutter engine lerping functions (from `flutter/engine/lib/ui/lerp.dart`) to lerp between 255 and 255 printing any values that are not 255 along the way.

**Expected results:** empty output - because lerping between 255 and 255 _should_ only return values of 255

**Actual results:** 
```
254 at t=0.05555555555555555
254 at t=0.047619047619047616
254 at t=0.04
254 at t=0.023255813953488372
254 at t=0.022222222222222223
254 at t=0.017857142857142856
254 at t=0.0136986301369863
254 at t=0.010526315789473684
```

## Proposed Solution
Lerping was broken by this [commit](https://github.com/flutter/engine/commit/784e6d746e9262cf31bc5add8f2ec1318a9d26eb)
In the [gist](https://gist.github.com/royzwerneman/a21701355f15a9f2cddff0aa39816415) linked above uncomment line 22 to see this reverted and correct behavior restored.

<details>
  <summary>Logs</summary>

actual crash that is logged when a non-opaque alpha value (244) results when lerping between theme colors:

```
════════ Exception caught by widgets library ═══════════════════════════════════════════════════════
The following assertion was thrown building WidgetsApp(dirty, state: _WidgetsAppState#7921a):
'package:flutter/src/widgets/title.dart': Failed assertion: line 26 pos 15: 'color != null && color.alpha == 0xFF': is not true.


Either the assertion indicates an error in the framework itself, or we should provide substantially more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new?template=BUG.md

When the exception was thrown, this was the stack: 
#2      new Title (package:flutter/src/widgets/title.dart:26:15)
#3      _WidgetsAppState.build (package:flutter/src/widgets/app.dart:1567:15)
#4      StatefulElement.build (package:flutter/src/widgets/framework.dart:4744:28)
#5      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4627:15)
#6      StatefulElement.performRebuild (package:flutter/src/widgets/framework.dart:4800:11)
...
════════════════════════════════════════════════════════════════════════════════════════════════════
```

flutter doctor -v

```
[✓] Flutter (Channel stable, 1.22.0, on Mac OS X 10.15.7 19H2, locale en-US)
    • Flutter version 1.22.0 at /Users/roy.zwerneman/development/flutter
    • Framework revision d408d302e2 (7 days ago), 2020-09-29 11:49:17 -0700
    • Engine revision 5babba6c4d
    • Dart version 2.10.0

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/roy.zwerneman/Library/Android/sdk
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = /Users/roy.zwerneman/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

 
[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 49.0.1
    • Dart plugin version 192.8052
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] VS Code (version 1.48.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.14.0

[✓] Connected device (1 available)
    • iPad (8th generation) (mobile) • 8466C310-622C-4862-A5C5-06F08595DE7B • ios • com.apple.CoreSimulator.SimRuntime.iOS-14-0 (simulator)

• No issues found!
```

</details>

**Comment by TahaTesser** [comment]
<details> 
<summary>code sample</summary> 

```dart
main() {
  // do lerp for 255 to 255
  for (var i = 1; i <= 99; i++) {
    double t = 1.0 / i;
    int x = _clampInt(_lerpInt(255, 255, t).toInt(), 0, 255);
    if (x != 255) print('${x} at t=${t}'); // print if the value is ever not 255
  }
}

// function copied from flutter engine lib/ui/lerp.dart
int _clampInt(int value, int min, int max) {
  assert(min <= max);
  if (value < min) {
    return min;
  } else if (value > max) {
    return max;
  } else {
    return value;
  }
}

// function copied from flutter engine lib/ui/lerp.dart
double _lerpInt(int a, int b, double t) {
//  return a + (b - a) * t; // this works well but was replaced with the below
  return a * (1.0 - t) + b * t;
}

```

</details> 

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel stable, 1.22.0, on Mac OS X 10.15.7 19H2, locale en-GB)
    • Flutter version 1.22.0 at /Users/tahatesser/Code/flutter_stable
    • Framework revision d408d302e2 (8 days ago), 2020-09-29 11:49:17 -0700
    • Engine revision 5babba6c4d
    • Dart version 2.10.0

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.10.0.rc.1

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 50.0.1
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.15.0

[✓] Connected device (1 available)
    • iPhone 11 (mobile) • AFF9AE1D-D6C7-46C4-9288-A98DB74D942B • ios •
      com.apple.CoreSimulator.SimRuntime.iOS-14-0 (simulator)

• No issues found!
```

</details> 

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel master, 1.23.0-8.0.pre.185, on Mac OS X 10.15.7 19H2
    x86_64, locale en-GB)
    • Flutter version 1.23.0-8.0.pre.185 at
      /Users/tahatesser/Code/flutter_master
    • Framework revision 6620f831e9 (8 hours ago), 2020-10-06 18:15:15 -0700
    • Engine revision 7e6191de07
    • Dart version 2.11.0 (build 2.11.0-186.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.10.0.rc.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 50.0.1
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.49.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.15.0

 
[✓] Connected device (4 available)            
    • iPhone 11 (mobile) • AFF9AE1D-D6C7-46C4-9288-A98DB74D942B • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-0 (simulator)
    • macOS (desktop)    • macos                                • darwin-x64
      • Mac OS X 10.15.7 19H2 x86_64
    • Web Server (web)   • web-server                           • web-javascript
      • Flutter Tools
    • Chrome (web)       • chrome                               • web-javascript
      • Google Chrome 85.0.4183.121

• No issues found!
```

</details>

**Comment by jason-simmons** [comment]
@cbracken should the `_lerpInt` part of https://github.com/flutter/engine/pull/20879 be reverted?

**Comment by cbracken** [comment]
Will put together a patch -- those patches fixed the exact same issue but for cases where there was a huge magnitude in difference between two values. I think the right answer is to add additional tests that cover this scenario and do a conditional check to determine which of the two forms to apply. I'll self-assign and write this up.

**Comment by cbracken** [comment]
This should be fixed by flutter/engine#22089. Added a test to ensure we don't regress this. No changes required on the web side since JS numerics are IEEE floats which behave correctly in this scenario, and which the original patch fixes.

@royzwerneman Thanks again for reporting this.

**Comment by cbracken** [comment]
I'll follow up sometime next week with the conditional logic in lerpDouble that I mentioned since, while it fixes a serious bug with values that differ significantly in magnitude, there are a few cases where we can get better precision in the final two decimals using the previous logic when we can verify it's safe.
